components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/main/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImpl.java

```diff
@@ -0,0 +1,992 @@
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.UserSharingConstants.ErrorMessage.ERROR_SKIP_SHARE;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.UserSharingConstants.ORGANIZATION;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.UserSharingConstants.USER_IDS;
+import static org.wso2.carbon.identity.organization.management.service.util.Utils.getOrganizationId;
+
+public class UserSharingPolicyHandlerServiceImpl implements UserSharingPolicyHandlerService {
+
+    private static final Log LOG = LogFactory.getLog(UserSharingPolicyHandlerServiceImpl.class);
+    private final UserIDResolver userIDResolver = new UserIDResolver();
+    private static ConcurrentLinkedQueue<String> errorMessages; //todo: populate this at last
+
+    @Override
+    public void populateSelectiveUserShare(SelectiveUserShareDO selectiveUserShareDO) throws UserShareMgtException {
+
+        LOG.debug("Came in user selective share");
+        validateUserShareInput(selectiveUserShareDO);
+        LOG.debug("Validated user selective share input");
+
+        //todo: run async from here
+        List<SelectiveUserShareOrgDetailsDO> organizations = selectiveUserShareDO.getOrganizations();
+        Map<String, UserCriteriaType> userCriteria = selectiveUserShareDO.getUserCriteria();
+
+        String sharingInitiatedOrgId = getOrganizationId();
+        List<String> immediateChildOrgs = getChildOrgsOfSharingInitiatedOrg(sharingInitiatedOrgId);
+
+        for (SelectiveUserShareOrgDetailsDO organization : organizations) {
+            if (immediateChildOrgs.contains(organization.getOrganizationId())) {
+                populateSelectiveUserShareByCriteria(organization, userCriteria);
+            } else {
+                throw new UserShareMgtClientException(ERROR_SKIP_SHARE);
+            }
+        }
+    }
+
+    private void populateSelectiveUserShareByCriteria(SelectiveUserShareOrgDetailsDO organization,
+                                                      Map<String, UserCriteriaType> userCriteria)
+            throws UserShareMgtException {
+
+        for (Map.Entry<String, UserCriteriaType> criterion : userCriteria.entrySet()) {
+            String criterionKey = criterion.getKey();
+            UserCriteriaType criterionValues = criterion.getValue();
+
+            switch (criterionKey) {
+                case USER_IDS:
+                    if (criterionValues instanceof UserIdList) {
+                        selectiveUserShareByUserIds((UserIdList) criterionValues, organization);
+                    } else {
+                        throw new UserShareMgtClientException(ERROR_CODE_USER_CRITERIA_INVALID);
+                    }
+                    break;
+                default:
+                    throw new UserShareMgtClientException(ERROR_CODE_USER_CRITERIA_INVALID);
+            }
+        }
```

Comment:
```suggestion
        for (Map.Entry<String, UserCriteriaType> criterion : userCriteria.entrySet()) {
        String criterionKey = criterion.getKey();
        UserCriteriaType criterionValues = criterion.getValue();

        if (USER_IDS.equals(criterionKey)) {
            if (criterionValues instanceof UserIdList) {
                selectiveUserShareByUserIds((UserIdList) criterionValues, organization);
            } else {
                throw new UserShareMgtClientException(ERROR_CODE_USER_CRITERIA_INVALID);
            }
        } else {
            throw new UserShareMgtClientException(ERROR_CODE_USER_CRITERIA_INVALID);
        }
    }
```

Can'y we use this if else instead of switch case here. Let's use switch case if we have more than 3 cases to handle

Reply:
@sahandilshan, I considered keeping the switch case to account for potential future criteria. What’s your opinion? Should we replace it with an if-else construct for now and revisit the switch case if more conditions are added later?

cc: @AnuradhaSK @ShanChathusanda93 @SujanSanjula96 @Yasasr1 @HasiniSama